src/pypromice/postprocess/get_bufr.py

```diff
@@ -5,287 +5,343 @@
-
-	# Iterate through csv files
-	for f in fpaths:
-		last_index = f.rfind('_')
-		first_index = f.rfind('/')
-		stid = f[first_index+1:last_index]
-		# stid = f.split('/')[-1].split('.csv')[0][:-5]
-
-		print('####### Processing {} #######'.format(stid))
-		if ('Roof' not in f) and (stid not in to_skip):
-		# if ('v3' not in f) and ('Roof' not in f) and (stid not in to_skip):
-			bufrname = stid + '.bufr'
-			print(f'Generating {bufrname} from {f}')
-
-			if (args.positions is True) and (stid not in positions_update_timestamp_only):
-				positions[stid] = {}
-				# Optionally include source flag columns, useful to indicate if position
-				# comes from current transmission, or older data. This could also be used
-				# to differentiate GPS from modem, but using the combine_first method for
-				# the modem positions currently prevents us from easily knowing which source
-				# was used.
-				# positions[stid]['lat_source'] = ''
-				# positions[stid]['lon_source'] = ''
-
-			# Read csv file
-			df1 = pd.read_csv(f, delimiter=',')
-			df1.set_index(pd.to_datetime(df1['time']), inplace=True)
-			df1.sort_index(inplace=True) # make sure we are time-sorted
-
-			# Check that the last valid index for all instantaneous values match
-			# Note: we cannot always use the single most-recent timestamp in the dataframe
-			# e.g. for 6-hr transmissions, *_u will have hourly data while *_i is nan
-			# Need to check for last valid (non-nan) index instead
-			lvi = {'t_i': df1['t_i'].last_valid_index(),
-				   'p_i': df1['p_i'].last_valid_index(),
-				   'rh_i': df1['rh_i'].last_valid_index(),
-				   'wspd_i': df1['wspd_i'].last_valid_index(),
-				   'wdir_i': df1['wdir_i'].last_valid_index()
-				   }
-
-			two_days_ago = datetime.utcnow() - timedelta(days=2)
-
-			if len(set(lvi.values())) != 1:
-				# instantaneous vars have different timestamps
-				recent = {}
-				for k,v in lvi.items():
-					if (v is not None) and (v >= two_days_ago):
-						recent[k] = v
-				if len(recent) == 0:
-					print('No recent instantaneous timestamps!')
-					no_recent_data.append(stid)
-					if args.positions is True:
-						df1_limited, positions = find_positions(df1, stid, args.time_limit, positions=positions)
-					continue
-				else:
-					# we have partial data, just use the most recent row
-					current_timestamp = max(recent.values())
-					# We will throw this obset down the line, and there is a final min_data_check
-					# to make sure we have minimum data requirements before writing to BUFR
-			else:
-				if all(i is None for i in lvi.values()) is True:
-					print('All instantaneous timestamps are None!')
-					no_valid_data.append(stid)
-					if args.positions is True:
-						df1_limited, positions = find_positions(df1, stid, args.time_limit, positions=positions)
-					continue
-				else:
-					# all values are present, with matching timestamps, so just use t_i
-					current_timestamp = df1['t_i'].last_valid_index()
```

Comment:
This section above was intended to find the `current_timestamp`, considering many different possible variations in the reported data and timestamps. This is ultimately intended for use in:

`if (current_timestamp > latest_timestamp) and (current_timestamp > two_days_ago):`

Make sure this time check is accounted for in the refactored version.

Reply:
This also allows us to update `positions`, even if the transmission does not have instantaneous values.

Reply:
@ladsmund if you follow the exact logic in this section, it should make sense (along with comments).... although I admit it is a bit confusing to follow.
https://github.com/GEUS-Glaciology-and-Climate/pypromice/blob/2ff4e4685ae66182afc32c690e4e437a40c43b6e/src/pypromice/postprocess/get_bufr.py#L140

Regarding partial data, in the existing code, this section:
```
				else:
					# we have partial data, just use the most recent row
					current_timestamp = max(recent.values())
					# We will throw this obset down the line, and there is a final min_data_check
					# to make sure we have minimum data requirements before writing to BUFR
```
This will keep only the most recent row/timestamp for a valid (non-null) instantaneous var. Then `current_timestamp` is used to index into the dataframe as:
```
					# limit to single most recent valid row (convert to series)
					s1_current = df1_limited.loc[current_timestamp]
```
So, for your example above, I think that is correct with `p_i: nan` and `t_i: -6`. In this case, `p_i` was already reported in the BUFR for the 10:00 processing run, and should not be submitted again with the 11:00 run. We want to only submit values that are associated with the timestamp they are reported.

In the case we have partial instantaneous data, we will check in `min_data_check` whether we have required vars to submit the BUFR. As currently implemented, we require lat, lon, elev, air temp, and pressure.

Reply:
To summarize, instead of selecting the latest valid _values_ from each variable, I think the idea is that we select using the last valid _timestamp_.

Reply:
As another example, we could have a case where some other variable (wind speed) is NaN, but air temp and pressure are still good. In that case, we don't care that wind speed is NaN, and using the last valid timestamp we will proceed with just air temp and pressure.

Reply:
👍 That make sense